feat(#1319): Add guide for Medic staff to start using Medic hosted EKS (#1320)
medic/cht-docs: content/en/contribute/code/core/deploy-on-eks.md

---
title: "Deploy CHT Core on Medic hosted EKS"
weight: 2
linkTitle: "Deploy to EKS"
description: >
    Setting up a cloud hosted deployment of CHT Core on Medic's AWS EKS infrastructure

> relatedContent: >
> 

Reviewer: Minor: Can this be removed?
```suggestion
```

5. Run `aws configure` and place appropriate access keys during prompts. Use `eu-west-2` region. It should look like this:

   ```
   $ aws configure

   AWS Access Key ID [None]: <ACCESS-KEY-HERE>
   AWS Secret Access Key [None]: <SECRET-HERE>
   Default region name [None]: eu-west-2
   Default output format [None]:
   ```
6. Run the Update Kubeconfig command, assuming username is `mrjones` and namespace is `mrjones-dev` - be sure to place these with yours: `aws eks update-kubeconfig --name mrjones-dev --profile mrjones --region eu-west-2`

## Autocomplete

Both `helm` and `kubectl` have autocomplete libraries you can add to your shell. For power users and beginners alike, this adds a huge amount of discoverability. This guide covers `zsh`, but `bash`, `fish` and `powershell` are supported as well:

1. Output `zsh` shell's setup to a file: 
   ```shell
   kubectl completion zsh > zsh.kubectl.autocomplete
   helm completion zsh > zsh.helm.autocomplete
   ```
  
2. Then source the two new files:
   ```shell
   source zsh.kubectl.autocomplete
   source zsh.helm.autocomplete
   ```

> 3. See [helm](https://helm.sh/docs/helm/helm_completion_bash/) and [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl-linux/#enable-shell-autocompletion) instructions for automatically loading these on every new session.

Reviewer: Does it make sense to move this section to be a subsection under **Command Line** above?

Author: yup - agreed!  also I can tighten them by combining the steps - see c4f8454d

## Starting and stopping (aka deleting)

> 1. Login with `eks-aws-mfa-login` script in the [infra repo](https://github.com/medic/medic-infrastructure/tree/master/terraform/aws/dev/eks/access): `./eks-aws-mfa-login USERNAME  TOTP_HERE`

Reviewer 2: Optional: It would be pretty helpful if this command and the below could be copied like in the `Autocomplete` section.

Author: Cool!  fixed in 0d8dc266

interestingly,  I don't see the "copy" button in the `Autocomplete` section - is this a macOS thing? I checked in FF, Chrome and Chromium on linux.

Reviewer 2: I'm on Linux @mrjones-plip  and firefox browser.
![image](https://github.com/medic/cht-docs/assets/30311933/c636d758-859c-4939-b4a0-45e335c41d10)

Author: huh - interesting! thanks for the update

(i run all sorts of "don't execute JS" plugins, likely it's me :sweat_smile: )

2. Ensure you're using dev EKS cluster: `kubectl config use-context arn:aws:eks:eu-west-2:720541322708:cluster/dev-cht-eks`

> 3. Add new `values.yaml` file - you can [copy this one](https://github.com/medic/medic-infrastructure/blob/master/terraform/aws/dev/cht-projects/alpha-dev-cht-deploy-values.yaml) and just change the `alpha-dev` values to `USERNAME-dev`

Author: ah - sure!  we can tell 'em "don't forget" as it doesn't actually matter where it is, but that they know where it is ;) 

fixed in 9a0a4d33

4. Use `uuidgen` to fill in the `secret` in `values.yaml`

> 5. Use a good passphrase (diceware!) to fill in `1password` in `values.yaml`

Author: ah - good catch! fixed in 7c9e3ea5

* list all resources: `./troubleshooting/list-all-resources mrjones-dev` 
* view logs, assuming `cht-couchdb-1` returned from prior command: `./troubleshooting/view-logs mrjones-dev cht-couchdb-1`
* describe deployment, assuming `cht-couchdb-1` returned from 1st command: `./troubleshooting/describe-deployment mrjones-dev cht-couchdb-1`
* list all deployments: `./troubleshooting/list-all-resources mrjones-dev`

### Getting shell

Sometimes you need to look at files and other key pieces of data that are not available with [the current](https://github.com/medic/cht-core/blob/master/scripts/deploy/troubleshooting/view-logs) `troubleshooting/view-logs` script.  In this case, getting an interactive shell on the pod can be helpful. 

1. First, get a list pods for your namespace: `kubectl -n NAMESPACE get pods`
2. After finding the pod you're interested, connect to the pod to get a shell: `kubectl -n NAMESPACE exec -it PODNAME/CONTAINERNAME -- /bin/bash`

### `invalid apiVersion` Error

If you get the error: 

> exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1" when running `kubectl version`

> You are using an outdated version of kubernetes api `client.authentication.k8s.io` which is not supported by your `kubectl` client. This can sometimes happen if aws cli is an older version, ensure you are running at least version `2` of aws cli by running: `aws --version`

Reviewer: Since the target environment is unknown I don't feel as sure of this
```suggestion
You might be using an version of kubernetes api `client.authentication.k8s.io` which is not supported by your `kubectl` client. This can sometimes happen in EKS clusters if aws cli is an older version, in most cases you need at least version `2` of aws cli. Check version by running: `aws --version` and note that version `2` *cannot* be installed through `pip` (See [Command Line](#command-line) section above for installation instructions)
```

---
title: "Deploy CHT Core on Medic hosted EKS"
weight: 2
linkTitle: "Deploy to EKS"
description: >
    Setting up a cloud hosted deployment of CHT Core on Medic's AWS EKS infrastructure
relatedContent: >
    core/overview/architecture
    core/overview/cht-sync
---

While not directly available to the public who might be doing CHT Core development, having Medic's process for using our [Amazon Elastic Kubernetes Service](https://docs.aws.amazon.com/eks/latest/userguide/what-is-eks.html) (AWS EKS) publicly documented will help Medic employees new to EKS.  As well, hopefully  external developers looking to re-use Medic tools and process to use EKS will find it helpful.

While these instructions assume you work at Medic and have access to private GitHub repositories, many of the tools are fully open source.

## Command Line Prerequisites 

> Before using this guide, be sure you have these tools installes:

Reviewer 3: ```suggestion
Before using this guide, be sure you have these tools installed:
```

* [awscli](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html): version `2` or newer
* [kubectl](https://kubernetes.io/docs/tasks/tools): Must be within one minor version of cluster. If cluster is `1.24.x`, use `1.23.x`, `1.24.x` or `1.25.x`.
* [helm](https://helm.sh/docs/intro/install/)
* [jq](https://jqlang.github.io/jq/download/)

## Autocomplete

Both `helm` and `kubectl` have autocomplete libraries you can add to your shell. For power users and beginners alike, this adds a huge amount of discoverability. This guide covers `zsh`, but `bash`, `fish` and `powershell` are supported as well:

1. Output `zsh` shell's setup to a file: 
   ```shell
   kubectl completion zsh > zsh.kubectl.autocomplete
   helm completion zsh > zsh.helm.autocomplete
   ```
  
2. Then source the two new files:
   ```shell
   source zsh.kubectl.autocomplete
   source zsh.helm.autocomplete
   ```

## Request permission

> 1. [Create a ticket](https://github.com/medic/medic-infrastructure/issues/new) to get your DNS and Namespace created for EKS, which  should match eachother. For me, `mrjones-dev` name space would match `mrjones.dev.medicmobile.org` DNS. The ticket should include requesting access to be granted EKS access. 

Reviewer 3: ```suggestion
1. [Create a ticket](https://github.com/medic/medic-infrastructure/issues/new) to get your DNS and Namespace created for EKS, which should match each other. As an example, a `USERNAME-dev` name space would match `USERNAME.dev.medicmobile.org` DNS. The ticket should include requesting EKS access to be granted. 
```

> 2. Once you get your AWS account invite  (see prior step), follow the [CLI setup guide](https://github.com/medic/medic-infrastructure/blob/master/terraform/aws/dev/eks/access/README.md).

Reviewer 3: ```suggestion
2. Once you get your AWS account invite (see prior step), follow the [CLI setup guide](https://github.com/medic/medic-infrastructure/blob/master/terraform/aws/dev/eks/access/README.md).
```

> **NB** - Security key (eg Yubikey) users need to add a TOTP MFA too! CLI requires the TOTP values (6-digit number) and security keys are not supported. Security keys can only be used on web logins.

Reviewer 3: ```suggestion
**NB** - Security key (e.g. Yubikey) users need to add a TOTP MFA (Time-based, One-Time Password Multi-Factor Authentication) too! CLI requires the TOTP values (6-digit number) and security keys are not supported. Security keys can only be used on web logins.
```

## Starting and stopping (aka deleting)

1. Login with `eks-aws-mfa-login` script in the [infra repo](https://github.com/medic/medic-infrastructure/tree/master/terraform/aws/dev/eks/access): `./eks-aws-mfa-login USERNAME  TOTP_HERE`
2. Ensure you're using dev EKS cluster. Replace `AWS_ACCOUNT_ID` with the real ID from SRE: `kubectl config use-context arn:aws:eks:eu-west-2:AWS_ACCOUNT_ID:cluster/dev-cht-eks`
3. Add new `values.yaml` file - you can [copy this one](https://github.com/medic/medic-infrastructure/blob/master/terraform/aws/dev/cht-projects/alpha-dev-cht-deploy-values.yaml) and just change the `alpha-dev` values to `USERNAME-dev`
4. Use `uuidgen` to fill in the `secret` in `values.yaml`

> 5. Use a good passphrase (diceware!) to fill in 1password1 in 1values.yaml1

Reviewer 3: Question: Is `1password1 in 1values.yaml1` correct? It looked like a lot of 1-s.

Author: ah - yes - good catch!  fixed in 7a7fb741
